### PR TITLE
planning: visualize open space trajectory in dreamview

### PR DESCRIPTION
**Pull request by deidaraho**

### CODE REVIEW

File: modules/planning/common/open_space_info.cc [hunk 1]
```diff
@@ -23,5 +23,9 @@
 namespace apollo {
 namespace planning {
 OpenSpaceInfo::OpenSpaceInfo() {}
+// load open space info to open space info debug
+void OpenSpaceInfo::record_debug() {
+    // TO-DO:
```
Comment (xiaoxq) [inline modules/planning/common/open_space_info.cc]: Indents. And Camel Case.

File: modules/planning/common/open_space_info.h [hunk 2]
```diff
@@ -334,6 +350,8 @@ class OpenSpaceInfo {
 
   std::pair<PublishableTrajectory, canbus::Chassis::GearPosition>
       publishable_trajectory_data_;
+
+  apollo::planning_internal::Debug* debug_;
```
Comment (xiaoxq) [inline modules/planning/common/open_space_info.h]: The ownership is unclear and dangerous. Use shared_ptr if you would like, though better to redesign.

File: modules/planning/on_lane_planning.cc [hunk 3]
```diff
@@ -392,6 +399,13 @@ Status OnLanePlanning::Plan(
         frame_->open_space_info().publishable_trajectory_data().second;
     publishable_trajectory.PopulateTrajectoryProtobuf(trajectory_pb);
     trajectory_pb->set_gear(publishable_trajectory_gear);
+
+    if (FLAGS_enable_record_debug) {
+      // call open space info load debug
+      if (FLAGS_export_chart) {
```
Comment (xiaoxq) [inline modules/planning/on_lane_planning.cc]: FLAGS_enable_record_debug && FLAGS_export_chart

File: modules/planning/on_lane_planning.cc [hunk 4]
```diff
@@ -486,5 +500,176 @@ bool OnLanePlanning::CheckPlanningConfig(const PlanningConfig& config) {
...
+    const ::google::protobuf::RepeatedPtrField<SpeedPlan>& speed_plans,
+    Chart* chart) {
+  chart->set_title("Speed Plan");
+  auto* options = chart->mutable_options();
+  options->mutable_x()->set_min(0.0);
+  options->mutable_x()->set_max(80.0);
+  options->mutable_x()->set_label_string("s (meter)");
+  options->mutable_y()->set_min(0.0);
+  options->mutable_y()->set_max(50.0);
+  options->mutable_y()->set_label_string("v (m/s)");
```
Comment (xiaoxq) [inline modules/planning/on_lane_planning.cc]: Repeated too many times. Add an anonymous helper function like PopulateChartOptions(x_min, x_max, x_label, y_min, y_max, y_label, *options) and reuse it.

File: modules/planning/common/open_space_info.cc (hunk 1 above)
Comment (deidaraho) [reply modules/planning/common/open_space_info.cc]: > Indents. And Camel Case.

fixed

File: modules/planning/on_lane_planning.cc (hunk 3 above)
Comment (deidaraho) [reply modules/planning/on_lane_planning.cc]: > FLAGS_enable_record_debug && FLAGS_export_chart

fixed

File: modules/planning/on_lane_planning.cc (hunk 4 above)
Comment (deidaraho) [reply modules/planning/on_lane_planning.cc]: > Repeated too many times. Add an anonymous helper function like PopulateChartOptions(x_min, x_max, x_label, y_min, y_max, y_label, *options) and reuse it.

fixed

File: modules/planning/common/open_space_info.h (hunk 2 above)
Comment (deidaraho) [reply modules/planning/common/open_space_info.h]: > The ownership is unclear and dangerous. Use shared_ptr if you would like, though better to redesign.

The debug pointer is passed by ADCtrajectory and we will modify the structure later.

File: modules/planning/common/open_space_info.cc [hunk 5]
```diff
@@ -23,5 +23,6 @@
 namespace apollo {
 namespace planning {
 OpenSpaceInfo::OpenSpaceInfo() {}
+// load open space info to open space info debug
```
Comment (xiaoxq) [inline modules/planning/common/open_space_info.cc]: This .cc is so dummy. You can define it as `= default;` in the header file.

File: modules/planning/common/open_space_info.h (hunk 2 above)
Comment (xiaoxq) [reply modules/planning/common/open_space_info.h]: Not good for ownership and lifecycle management. You can just use the debug_instance_ and copy to the external ADCtrajectory in the end. Cleaner and safer.

File: modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_provider.cc [hunk 6]
```diff
@@ -324,5 +337,15 @@ void OpenSpaceTrajectoryProvider::ReuseLastFrameResult(
   frame_->mutable_open_space_info()->set_open_space_provider_success(true);
 }
 
+void OpenSpaceTrajectoryProvider::ReuseLastFrameDebug(
+    const Frame* last_frame) {
+  // reuse last frame's instance
+  auto* debug_ptr = frame_->mutable_open_space_info()->mutable_debug_instance();
+  debug_ptr->mutable_planning_data()->mutable_open_space()->MergeFrom(
+    last_frame->open_space_info().debug_instance().
+      planning_data().open_space());
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_provider.cc]: Wrong indents. Setup and run scripts/clang-format.sh to clean.

File: modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_optimizer.cc [hunk 7]
```diff
@@ -196,6 +206,130 @@ common::Status OpenSpaceTrajectoryOptimizer::Plan(
...
+  open_space_debug_.add_xy_boundary(XYbounds[2]);
+  open_space_debug_.add_xy_boundary(XYbounds[3]);
+
+  // load obstacles
+  for (const auto& obstacle_vertices : obstacles_vertices_vec) {
+    auto* obstacle_ptr = open_space_debug_.add_obstacles();
+    size_t vertices_size = obstacle_vertices.size();
+    for (size_t i = 0; i < vertices_size; ++i) {
+      obstacle_ptr->add_vertices_x_coords(obstacle_vertices[i].x());
+      obstacle_ptr->add_vertices_y_coords(obstacle_vertices[i].y());
```
Comment (xiaoxq) [inline modules/planning/tasks/optimizers/open_space_trajectory_generation/open_space_trajectory_optimizer.cc]: Loop with `for (const auto& vertex : obstacle_vertices)`.

File: modules/planning/common/open_space_info.cc (hunk 5 above)
Comment (deidaraho) [reply modules/planning/common/open_space_info.cc]: > This .cc is so dummy. You can define it as `= default;` in the header file.

fixed

**Pull request merged**